torchdata/datapipes/iter/transform/callable.py

```diff
@@ -585,3 +592,156 @@ def __new__(
+        source_datapipe: Source IterDataPipe
+        fn: The function to be applied to each element within batch of data
+        batch_size: The size of batch to be aggregated from ``source_datapipe``
+        input_col: Index or indices of data which ``fn`` is applied, such as:
+            - ``None`` as default to apply ``fn`` to the data directly.
+            - Integer(s) is used for list/tuple.
+            - Key(s) is used for dict.
+        output_col: Index of data where result of ``fn`` is placed. ``output_col`` can be specified
+            only when ``input_col`` is not ``None``
+            - ``None`` as default to replace the index that ``input_col`` specified; For ``input_col`` with
+              multiple indices, the left-most one is used, and other indices will be removed.
+            - Integer is used for list/tuple. ``-1`` represents to append result at the end.
+            - Key is used for dict. New key is acceptable.
+        max_workers: Maximum num of threads to execute function calls. (Default value: None)
+
+    Note:
+         For more information about ``max_workers`` and please refer to: https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor
+
+    Example:
+
+    .. testsetup::
+
+        from torchdata.datapipes.iter import IterableWrapper
+        import requests
+        import time
+        from unittest.mock import MagicMock
+
+        requests.get = MagicMock()
+        urls = []
+
+    .. testcode::
+
+        def mul_ten(x):
+            time.sleep(0.1)
+            return x * 10
+        dp = IterableWrapper(range(50))
+        dp = dp.thread_map_batches(mul_ten, 16)
+        print(list(dp))
+
+    .. testoutput::
+
+        [0, 10, 20, 30, ...]
+
+    .. testcode::
+
+        dp = IterableWrapper([(i, i) for i in range(50)])
+        dp = dp.thread_map_batches(mul_ten, 16, input_col=1)
+        print(list(dp))
+
+    .. testoutput::
+
+        [(0, 0), (1, 10), (2, 20), (3, 30), ...]
+
+    .. testcode::
+
+        dp = IterableWrapper([(i, i) for i in range(50)])
+        dp = dp.thread_map_batches(mul_ten, 16, input_col=1, output_col=-1)
+        print(list(dp))
+
+    .. testoutput::
+
+        [(0, 0, 0), (1, 1, 10), (2, 2, 20), (3, 3, 30), ...]
+
+    .. testcode::
+
+        # fetching html from remote
+        def fetch_html(url: str, **kwargs):
+            r = requests.get(url, **kwargs)
+            r.raise_for_status()
+            return r.content
+        dp = IterableWrapper(urls)
+        dp = dp.thread_map_batches(fetch_html, 16)
+
+    """
+
+    def __new__(
+        cls,
+        source_datapipe,
+        fn: Callable,
+        batch_size: int,
+        input_col=None,
+        output_col=None,
+        max_workers: Optional[int] = None,
+        **threadpool_kwargs,
+    ):
+        dp = source_datapipe.batch(batch_size)
+        dp = _BatchThreadPoolMapperIterDataPipe(dp, fn, input_col, output_col, max_workers, **threadpool_kwargs)
```

Comment:
I think it's fine as you said, a `prefetch` would help this problem. The main reason we do it rather yielding data whenever one data has been processed by `map` is that we need to guarantee the order of data.

Reply:
Should we expose the `buffer_size` or set it to some fixed value (2 to start working on only the next batch or default_value of `Prefetcher`)?
We probably should also make recommendations for `batch_size`. 
For both we probably need to do some performance tests.

Reply:
IMHO, prefetching 1 element should be sufficient as requesting the first element would lead to launching all threads for the next batch.

Reply:
Ah, I thought about adding Prefetcher before `flatmap`, but yeah the next batch is probably enough.

Reply:
I have done a small test running in the background:
```python
def wait_a_bit(x):
    time.sleep(random.random() * 3)
    return x * 10

dp = IterableWrapper(range(10000))
print("threads, batch_size, time until 10000 ")
for batch_size in map(lambda x : pow(2,x), reversed(range(7,13))):
    start = time.time()
    for _ in dp.threadpool_map(wait_a_bit,batch_size=batch_size, max_workers=128):
        pass
    end = time.time()
    print(f"{threads}, {batch_size}, {end - start}")
```
without `prefetching`: 
| threads | batch_size | time until 10000 |
|---------|------------|------------------|
| 128     | 4096       | 122.96           |
| 128     | 2048       | 126.84           |
| 128     | 1024       | 135.94           |
| 128     | 512        | 154.29           |
| 128     | 256        | 192.07           |
| 128     | 128        | 236.32           |

with `prefetching` and `buffer_size=2`:
| threads | batch_size | time until 10000   |
|---------|------------|--------------------|
| 128     | 4096       | 117.93 |
| 128     | 2048       | 120.48  |
| 128     | 1024       | 120.39 |
| 128     | 512        | 119.98    |
| 128     | 256        | 120.47 |
| 128     | 128        | 119.05 |

The theoretical minimum is at around 117s. As you can see adding a `Prefetcher` is necessary. WIth `buffer_size = 1` we get worse results if the `batch_size` is the same (or smaller) than the number of threads but similar results if it is larger. Therefore I think `buffer_size=2` is the right choice.

Reply:
I agree `Prefetch` is necessary. I was incorrect about using 1 would be sufficient. I guess the more complex of the map function, the larger `buffer_size` is needed.
On top of `time until 10000`, it might be good to have the time to get the first element.

Reply:
Interistingly this also depends on the use of `Prefetcher` and `batch_size` but also on `buffer_size`

without prefetching:
| threads | batch_size | time until first |
|---------|------------|------------------|
| 128     | 4096       | 4.61             |
| 128     | 2048       | 4.53             |
| 128     | 1024       | 4.86             |
| 128     | 512        | 3.34             |
| 128     | 256        | 3.20             |
| 128     | 128        | 3.01             |
| 128     | 64         | 3.01             |
| 128     | 32         | 2.93             |
| 32      | 128        | 4.48             |
| 32      | 64         | 3.36             |
| 32      | 32         | 2.99             |

with `buffer_size=2`:
| threads | batch_size | time until first |
|---------|------------|------------------|
| 128     | 2048       | 37.62            |
| 128     | 1024       | 28.21            |
| 128     | 512        | 16.08            |
| 128     | 256        | 9.45             |
| 128     | 128        | 6.06             |
| 128     | 64         | 3.91             |
| 128     | 32         | 3.00             |
| 32      | 128        | 17.34 |
| 32      | 64         | 10.44  |
| 32      | 32         | 5.86  |

with `buffer_size=1`

| threads | batch_size | time until first |
|---------|------------|------------------|
| 128     | 2048       | 27.07            |
| 128     | 1024       | 16.39            |
| 128     | 512        | 10.42            |
| 128     | 256        | 6.88             |
| 128     | 128        | 4.90             |
| 128     | 64         | 3.02             |
| 128     | 32         | 3.03             |
| 32      | 128        | 10.35 |
| 32      | 64         | 6.78  |
| 32      | 32         | 4.64 |

Without prefetching (AND without specifying an `input_col` !) we make use of the fact that the result of `executor.map` is an iterator and not the whole batch. 

So it really is a tradeoff. Maybe add an argument to enable/disable prefetching depending on whats more important and set it to `True` by default

Reply:
Yeah, I agree that it's just tradeoff that users need to tune based on their own requirements. 
How about we remove `prefetch` from the factory DataPipe? And, just give them suggestions in the documents. Two major reasons:
- Different scenarios require different `buffer_size`. Users can always do it by themselves to attach a `prefetch`.
- If `MultiProcessingReadingService` is used, a `prefetch` is attached automatically, which might help the result.

Reply:
Adding a `Prefetch `after `flatmap` doesn't help improve performance as the next batch is only started once all elements from previous batch have been yielded.

Instead I rewrote ThreadPoolMapper to no longer use batches but schedule only a finite number of tasks at any time. This new implementation is better in every way:
If we e.g. choose `scheduled_tasks = 2 * threads` (64 threads, 128 scheduled tasks)

- First element is yielded relatively fast (4.78s)
- throughput (time until 5000) is still fast (119s)
- Access to `len` from previous dp
- Less dps in the graph
- Elements are still returned in correct order

Higher value of `scheduled_tasks` will lead to longer waiting period until first element is yielded as tasks are executed out of order.

Reply:
> Instead I rewrote ThreadPoolMapper to no longer use batches but schedule only a finite number of tasks at any time. This new implementation is better in almost every way:

It sounds reasonable to me. I will take a deeper look at your PR later today. For `asyncio`, I might need to revamp it as well. And, `asyncio` has another problem with nested asyncio enabled.